### PR TITLE
Gridmask Gpu

### PR DESCRIPTION
**Pull request by PawelA**
#### Why we need this PR?
*Pick one, remove the rest*
- It adds gpu version of gridmask operator & kernel

#### What happened in this PR?
*Fill relevant points, put NA otherwise. Replace anything inside []*
 - What solution was applied:
     Added gpu operator and kernel
 - Affected modules and functionalities:
     - Gridmask operator header - added class for gpu
     - Added gpu operator and kernel
 - Key points relevant for the review:
     Kernel efficiency
 - Validation and testing:
     Extended cpu tests to include gpu
 - Documentation (including examples):
     Already present in cpu operator

**JIRA TASK**: NA

### CODE REVIEW

File: dali/kernels/mask/grid_mask_gpu.h [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+  int x = 32 * blockIdx.x + threadIdx.x;
+  int y = 32 * blockIdx.y + threadIdx.y;
+  if (x >= args.width || y >= args.height)
+    return;
+
+  int off = args.channels * (x + y * args.width);
+  float fx = -args.sx + x * args.ca - y * args.sa;
+  float fy = -args.sy + x * args.sa + y * args.ca;
+  float m = (fx - floor(fx) >= args.ratio) || (fy - floor(fy) >= args.ratio);
+  for (int i = 0; i < args.channels; i++)
```
Comment (JanuszL) [inline dali/kernels/mask/grid_mask_gpu.h:43]: I would make `channels` a template argument and have a specialization for 1, 3 and 4, and generic loop for any other number of channels. You can check `dali/kernels/imgproc/flip_gpu.cuh` as an example.

File: dali/kernels/mask/grid_mask_gpu.h [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+template <typename T>
+__global__ void GridMaskKernel(T *output, const T *input, const Args args) {
+  int x = 32 * blockIdx.x + threadIdx.x;
+  int y = 32 * blockIdx.y + threadIdx.y;
+  if (x >= args.width || y >= args.height)
+    return;
+
+  int off = args.channels * (x + y * args.width);
+  float fx = -args.sx + x * args.ca - y * args.sa;
```
Comment (JanuszL) [inline dali/kernels/mask/grid_mask_gpu.h:40]: You can consider using fma instruction.

File: dali/operators/image/mask/grid_mask.cu [hunk 3]
```diff
@@ -0,0 +1,61 @@
+// Copyright (c) 2019, NVIDIA CORPORATION. All rights reserved.
```
Comment (JanuszL) [inline dali/operators/image/mask/grid_mask.cu:1]: I think you add a GPU variant in the relevant operator test.

File: dali/kernels/mask/grid_mask_gpu.h [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+namespace detail {
+
+struct Args {
+  int width, height, channels;
+  float ca, sa, sx, sy, ratio;
+};
+
+template <typename T>
+__global__ void GridMaskKernel(T *output, const T *input, const Args args) {
```
Comment (mzient) [inline dali/kernels/mask/grid_mask_gpu.h:33]: This is a per-sample kernel. We generally prefer batch processing.

File: dali/kernels/mask/grid_mask_gpu.h [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+  int x = 32 * blockIdx.x + threadIdx.x;
+  int y = 32 * blockIdx.y + threadIdx.y;
```
Comment (mzient) [inline dali/kernels/mask/grid_mask_gpu.h:35]: While horizontal warp-sized blocks are a good idea, the vertical size should be configurable and possibly tuned. General experience suggests that 256 or 512 threads per block yields better performance - in any case, this should be tuned.
Having said all that, the very first thing to address is batch processing.

File: dali/kernels/mask/grid_mask_gpu.h (hunk 4 above)
Comment (mzient) [reply dali/kernels/mask/grid_mask_gpu.h:33]: Look at block processing in other kernels - there are some utilities for that purpose - namely, BlockSetup.

File: dali/kernels/mask/grid_mask_gpu.h (hunk 2 above)
Comment (mzient) [reply dali/kernels/mask/grid_mask_gpu.h:40]: :+1: that's two FMAs

File: dali/kernels/mask/grid_mask_gpu.h [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+  if (x >= args.width || y >= args.height)
+    return;
+
+  int off = args.channels * (x + y * args.width);
+  float fx = -args.sx + x * args.ca - y * args.sa;
+  float fy = -args.sy + x * args.sa + y * args.ca;
+  float m = (fx - floor(fx) >= args.ratio) || (fy - floor(fy) >= args.ratio);
```
Comment (mzient) [inline dali/kernels/mask/grid_mask_gpu.h:42]: That's terribly inefficient - you don't have to multiply - you don't even have to read the input for pixels that are zeroed-out.
```suggestion
  if (fx - floor(fx) >= args.ratio) || (fy - floor(fy) >= args.ratio) {
    output[off + i] = input[off + i];
  } else {
    output[off + i] = 0;
  }
```

**Review by mzient — CHANGES_REQUESTED** [review]
Batch processing is required.

File: dali/kernels/mask/grid_mask_gpu.h (hunk 1 above)
Comment (PawelA) [reply dali/kernels/mask/grid_mask_gpu.h]: done

File: dali/kernels/mask/grid_mask_gpu.h (hunk 2 above)
Comment (PawelA) [reply dali/kernels/mask/grid_mask_gpu.h]: done

File: dali/kernels/mask/grid_mask_gpu.h (hunk 4 above)
Comment (PawelA) [reply dali/kernels/mask/grid_mask_gpu.h]: done

File: dali/kernels/mask/grid_mask_gpu.h (hunk 6 above)
Comment (PawelA) [reply dali/kernels/mask/grid_mask_gpu.h]: done

File: dali/kernels/mask/grid_mask_gpu.h (hunk 5 above)
Comment (PawelA) [reply dali/kernels/mask/grid_mask_gpu.h]: this is now handled by BlockSetup, but I'm not sure if I'm using it efficiently

File: dali/operators/image/mask/grid_mask.cu [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <vector>
+#include "dali/core/static_switch.h"
+#include "dali/operators/image/mask/grid_mask.h"
+#include "dali/kernels/mask/grid_mask_gpu.h"
+
+#define TYPES (float, uint8_t)
```
Comment (mzient) [inline dali/operators/image/mask/grid_mask.cu:20]: For the sake of potential collision with some library macro called TYPES:
```suggestion
#define GRID_MASK_TYPES (float, uint8_t)
```

File: dali/test/python/test_operator_gridmask.py [hunk 8]
```diff
@@ -24,24 +24,26 @@
...
-  pipe = Pipeline(batch_size, 4, None)
+def get_random_pipeline(device, batch_size):
+  pipe = Pipeline(batch_size, 4, 0)
   with pipe:
     input, _ = fn.file_reader(file_root=img_dir)
     decoded = fn.image_decoder(input, device='cpu', output_type=types.RGB)
+    decoded = decoded.gpu() if device == 'gpu' else decoded
     tile = fn.cast(fn.uniform(range=(50, 200), shape=[1]), dtype=types.INT32)
     ratio = fn.uniform(range=(0.3, 0.7), shape=[1])
     angle = fn.uniform(range=(-math.pi, math.pi), shape=[1])
```
Comment (mzient) [inline dali/test/python/test_operator_gridmask.py:45]: These should be scalars. If they're not, please change them in the operator.

**Review by mzient — APPROVED** [review]
Only very minor remarks. This is a beautifully written PR, real pleasure to look at.

**Comment by JanuszL** [comment]
!build

**Comment by JanuszL** [comment]
It looks like clang complains about:
```
/opt/dali/dali/kernels/mask/grid_mask_gpu.h:115:7: note: in instantiation of function template specialization 'dali::kernels::GridMaskKernel<float, 1>' requested here
      GridMaskKernel<Type, C><<<grid, block, 0, ctx.gpu.stream>>>(samples_gpu, blocks_gpu);
      ^
/opt/dali/dali/kernels/kernel_manager.h:249:24: note: in instantiation of member function 'dali::kernels::GridMaskGpu<float>::Run' requested here
    inst.get<Kernel>().Run(context, std::forward<OutInArgs>(out_in_args)...);
                       ^
/opt/dali/dali/kernels/kernel_manager.h:221:5: note: in instantiation of function template specialization 'dali::kernels::KernelManager::Run<dali::kernels::GridMaskGpu<float>, dali::TensorListView<dali::StorageGPU, float, 3> &, dali::TensorListView<dali::StorageGPU, const float, 3> &, std::vector<int, std::allocator<int> > &, std::vector<float, std::allocator<float> > &, std::vector<float, std::allocator<float> > &, std::vector<float, std::allocator<float> > &, std::vector<float, std::allocator<float> > &>' requested here
    Run<Kernel>(sa, instance_idx, context, std::forward<OutInArgs>(out_in_args)...);
    ^
/opt/dali/dali/operators/image/mask/grid_mask.cu:56:27: note: in instantiation of function template specialization 'dali::kernels::KernelManager::Run<dali::kernels::GridMaskGpu<float>, dali::TensorListView<dali::StorageGPU, float, 3> &, dali::TensorListView<dali::StorageGPU, const float, 3> &, std::vector<int, std::allocator<int> > &, std::vector<float, std::allocator<float> > &, std::vector<float, std::allocator<float> > &, std::vector<float, std::allocator<float> > &, std::vector<float, std::allocator<float> > &>' requested here
          kernel_manager_.Run<Kernel>(ws.thread_idx(), 0, ctx, out_view, in_view,
                          ^
In file included from /opt/dali/dali/operators/image/mask/grid_mask.cu:18:
/opt/dali/dali/kernels/mask/grid_mask_gpu.h:52:27: error: comparison of integers of different signs: 'int' and 'unsigned int' [-Werror,-Wsign-compare]
        for (int i = 0; i < (C ? C : sample.channels); i++)
                        ~ ^  ~~~~~~~~~~~~~~~~~~~~~~~
/opt/dali/dali/kernels/mask/grid_mask_gpu.h:49:27: error: comparison of integers of different signs: 'int' and 'unsigned int' [-Werror,-Wsign-compare]
        for (int i = 0; i < (C ? C : sample.channels); i++)
```

**Comment by JanuszL** [comment]
To run clang build you can:
```
EXTRA_CMAKE_OPTIONS="-DCMAKE_CXX_COMPILER=clang++ -DCMAKE_C_COMPILER=clang -DDALI_CLANG_ONLY=ON" CUDA_VERSION=100  ./build.sh
```

**Comment by PawelA** [comment]
I couldn't get the clang build to work. I also tried compiling grid_mask.cu with clang by hand to see the warning, but no luck, so I'm not certain I fixed the build fail.

**Comment by JanuszL** [comment]
!build

File: dali/test/python/test_operator_gridmask.py [hunk 9]
```diff
@@ -77,27 +79,33 @@ def check(result, input, tile, ratio, angle):
 
 def test_cpu_vs_cv():
```
[the remark this answers is not in the document]
Comment (PawelA) [reply dali/test/python/test_operator_gridmask.py]: done

File: dali/operators/image/mask/grid_mask.cu [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+namespace dali {
+
+bool GridMaskGpu::SetupImpl(std::vector<OutputDesc> &output_desc,
+                            const workspace_t<GPUBackend> &ws) {
+  const auto &input = ws.template InputRef<GPUBackend>(0);
+  const auto &output = ws.template OutputRef<GPUBackend>(0);
+  output_desc.resize(1);
+  GetArguments(ws);
+  output_desc[0] = {input.shape(), input.type()};
+  kernel_manager_.Resize(num_threads_, max_batch_size_);
```
[the remark this answers is not in the document]
Comment (PawelA) [reply dali/operators/image/mask/grid_mask.cu]: I changed it and it still works

**Comment by awolant** [comment]
!build

**Review by awolant — APPROVED** [review]

**Pull request merged**